### ISSUE TITLE
Write error in DFU Mode Ipad 3,2

### ISSUE THREAD

**Issue by madmanx2** [LukeZGD/32bit-OTA-Downgrader#13]
Hi, 
First of all thank you for coming up with this toolkit... much appreciated!  I'm running Lubuntu 18.04 live USB.  Ipad is JB via Phoenix.  Ipad has plenty of free space as it has been wiped and set up a new device prior to JB.  Everything goes fine with Downgrader until it gets to the part right after I hit the home button and the IPSW is trying to extract.  I get the write error below.  This repeats for several other files if I choose continue and ultimately fails later on in the process.  This appears to be the point at where the failure starts though.  Is the write error potentially due to Linux on my liveinstall not having enough RAM or is it referring to writing to the Ipad?  Not sure what to do next.
Thanks!

[Log] Found device in DFU mode.
[Log] Extracting IPSW...
iPad3,2_8.4.1_12H321_Restore/058-24297-023.dmg:  write error (disk full?).  Continue? (y/n/^C) 

******* 32bit-OTA-Downgrader *******
    Downgrade script by LukeZGD     

Main Menu

HardwareModel: J2AP
ProductType: iPad3,2
ProductVersion: 9.3.6
UniqueChipID (ECID): 2487659774762

[Input] Select an option:
1) Downgrade device		 4) (Re-)Install Dependencies
2) Save OTA blobs		 5) Exit
3) Just put device in kDFU mode
#? 1
[Input] Select iOS version:
1) iOS 8.4.1
2) Other
3) Back
#? 1
iOS 8.4.1 Downgrade
[Log] Downloading firmware keys...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   135  100   135    0     0    401      0 --:--:-- --:--:-- --:--:--   400
100 76459    0 76459    0     0   109k      0 --:--:-- --:--:-- --:--:--  109k
[Log] Saving 8.4.1 blobs with tsschecker...
Version: 7d267698cb16ab4699fa9cba20783ee041ac999e - 212
[TSSC] manually specified ecid to use, parsed "2487659774762" to dec:2487659774762 hex:2433413c32a
[TSSC] opening resources/manifests/BuildManifest_iPad3,2_8.4.1.plist
[WARNING] [TSSC] could not get id0 for installType=Erase. Using fallback installType=Update since user did not specify installType manually
[TSSR] Request URL set to https://gs.apple.com/TSS/controller?action=2
[TSSR] Sending TSS request attempt 1... success
Saved shsh blobs!

iOS 8.4.1 for device iPad3,2 IS being signed!
[Log] Verifying IPSW...
[Log] Extracting iBSS from IPSW...
Archive:  iPad3,2_8.4.1_12H321_Restore.ipsw
  inflating: tmp/iBSS.j2.RELEASE.dfu  
[Log] Decrypting iBSS...
IV = 32fcd912cb9a472ef2a6db72596ae01c
Key = 076720d5a07e8011bdda6f6eafaf4845b40a441615cd1d7c1a9cca438ce7db17
/home/tihm/odysseusOTA/xpwn/ipsw-patch/img3.c:createAbstractFileFromImg3:643: a40b3d8f62945a12480a15b533acbe1a7b14e51fe34e03e06aabf337aa7cfc1bd89630f75f314d2060af8669e9c18a99
/home/tihm/odysseusOTA/xpwn/ipsw-patch/img3.c:createAbstractFileFromImg3:643: a40b3d8f62945a12480a15b533acbe1a7b14e51fe34e03e06aabf337aa7cfc1bd89630f75f314d2060af8669e9c18a99
/home/tihm/odysseusOTA/xpwn/ipsw-patch/img3.c:createAbstractFileFromImg3:643: a40b3d8f62945a12480a15b533acbe1a7b14e51fe34e03e06aabf337aa7cfc1bd89630f75f314d2060af8669e9c18a99
1218+1 records in
1218+1 records out
77980 bytes (78 kB, 76 KiB) copied, 0.00482322 s, 16.2 MB/s
[Log] Patching iBSS...
Make sure SSH is installed and working on the device!
Please enter Wi-Fi IP address of device for SSH connection
[Input] IP Address: 192.168.1.131
[Log] Coonecting to device via SSH... Please enter root password when prompted (default is 'alpine')
[Log] Copying stuff to device...
root@192.168.1.131's password: 
kloader                                       100%   51KB   1.3MB/s   00:00    
pwnediBSS                                     100%   76KB   1.8MB/s   00:00    
[Log] Entering kDFU mode...

Press home/power button once when screen goes black on the device
[Log] Finding device in DFU mode...
root@192.168.1.131's password: 
[Log] Found device in DFU mode.
[Log] Extracting IPSW...
iPad3,2_8.4.1_12H321_Restore/058-24297-023.dmg:  write error (disk full?).  Continue? (y/n/^C) ^C[Log] Preparing for futurerestore (starting local server)...
[Log] Proceeding to futurerestore...
Version: b99eb8140d8e6c23f34e950102bb79e61c72384d - 152
Libipatcher Version: f32e41d850f51448bd6c588ead9c7d6455733f3c - 44
Odysseus Support: yes
[INFO] 32bit device detected
futurerestore init done
[Error] failed to load apticket at 2487659774762_iPad3,2_8.4.1-12H321_770d60607ff3b40ab72ffbc6bf8bfeddd5a46d44.shsh2
[Error] Fail code=-9
Failed with errorcode=-9

[Log] futurerestore done!
You can choose to retry if futurerestore failed on downloading baseband or for some other reason
[Input] Retry? (y/N) Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/usr/lib/python3.6/http/server.py", line 1211, in <module>
    test(HandlerClass=handler_class, port=args.port, bind=args.bind)
  File "/usr/lib/python3.6/http/server.py", line 1185, in test
    with ServerClass(server_address, HandlerClass) as httpd:
  File "/usr/lib/python3.6/socketserver.py", line 456, in __init__
    self.server_bind()
  File "/usr/lib/python3.6/http/server.py", line 136, in server_bind
    socketserver.TCPServer.server_bind(self)
  File "/usr/lib/python3.6/socketserver.py", line 470, in server_bind
    self.socket.bind(self.server_address)
OSError: [Errno 98] Address already in use

**Comment by LukeZGD** [comment]
How much RAM does your system have? This is most likely a problem only in live USB's and wouldn't be a problem in a full install. I also tested this only on a system with 8GB of RAM

**Comment by madmanx2** [comment]
My laptop has 6GB of ram.  It dual boots Windows 10 and Linux Mint  19.1.  I initially tried it in Mint, but couldn't get it to work - unsure if I chose the correct distro similar to Mint and also wasn't sure if it would install the proper dependencies, so I did the live USB install.  I'll launch it on on of my systems with more ram or setup a dual boot and give it another try and report back.

**Comment by LukeZGD** [comment]
Linux Mint 19.1 is based on Bionic, so Ubuntu Bionic dependencies should work on that, but I haved tested it myself yet

**Comment by LukeZGD** [comment]
I tested on Manjaro full install and it worked fine as expected

As for live USB, do not use the home folder, you may download and run the script from an external HDD or another USB drive with at least 8 GB free space.. I tested this with a Manjaro live USB and it worked fine as well

**Comment by madmanx2** [comment]
Thanks for all  your help!  I created a Manjaro live USB and followed your advice about not installing on the flash USB home, instead pointing it to the home downloads folder on the Mint partition.  Ran and executed great... getting past the point where the write error occurred and downgraded perfectly.  Thanks again for your support and response in working through the issue.  I'll have to try out Manjaro a bit... looks pretty good.
